easybuild/tools/module_naming_scheme/generation_mns.py

```diff
@@ -0,0 +1,164 @@
+        ConfigurationVariables()._FrozenDict__dict['module_naming_scheme'] = 'EasyBuildMNS'
+
+        # map generations on toolchains
+        for generation in self.generations:
+            for tc in get_toolchain_hierarchy({'name': 'foss', 'version': generation}):
+                self.lookup_table[(tc['name'], tc['version'])] = generation
+            # include (foss, <generation>) as a toolchain aswell
+            self.lookup_table[('foss', generation)] = generation
+
+        # Force config to point to other MNS
+        ConfigurationVariables()._FrozenDict__dict['module_naming_scheme'] = 'GenerationModuleNamingScheme'
+
+        # users can provide custom generation-toolchain mapping through a file
+        path = os.environ.get(GMNS_ENV)
+        if path:
+            if not os.path.isfile(path):
+                msg = "value of ENV {} ({}) should be a valid filepath"
+                raise EasyBuildError(msg.format(GMNS_ENV, path))
+            with open(path, 'r') as hc_lookup:
+                try:
+                    hc_lookup_data = json.loads(hc_lookup.read())
+                except json.decoder.JSONDecodeError:
+                    raise EasyBuildError("{} can't be decoded as json".format(path))
+                if not isinstance(hc_lookup_data, dict):
+                    raise EasyBuildError("{} should contain a dict".format(path))
+                if not set(hc_lookup_data.keys()) <= set(self.generations):
+                    raise EasyBuildError("Keys of {} should be generations".format(path))
+                for generation, toolchains in hc_lookup_data.items():
+                    if not isinstance(toolchains, list):
+                        raise EasyBuildError("Values of {} should be lists".format(path))
+                    for tc in toolchains:
+                        if not isinstance(tc, dict):
+                            msg = "Toolchains in {} should be of type dict"
+                            raise EasyBuildError(msg.format(path))
+                        if set(tc.keys()) != {'name', 'version'}:
+                            msg = "Toolchains in {} should have two keys ('name', 'version')"
+                            raise EasyBuildError(msg.format(path))
+                        self.lookup_table[(tc['name'], tc['version'])] = generation
+
+    def det_full_module_name(self, ec):
+        """
+        Determine full module name, relative to the top of the module path.
+        Examples: General/GCC/4.8.3, Releases/2018b/OpenMPI/1.6.5
+        """
+        return os.path.join(self.det_module_subdir(ec), self.det_short_module_name(ec))
+
+    def det_short_module_name(self, ec):
+        """
+        Determine short module name, i.e. the name under which modules will be exposed to users.
+        Examples: GCC/4.8.3, OpenMPI/1.6.5, OpenBLAS/0.2.9, HPL/2.1, Python/2.7.5
+        """
+        return os.path.join(ec['name'], self.det_full_version(ec))
```

Comment:
I don't believe that you can drop the toolchain/version suffix here without exposing yourself to lots of problems. How does this take into account that you can have, for example, OpenMPI with other compilers? Looking at a concrete example:
[OpenMPI-4.0.3-GCC-9.3.0.eb](https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.0.3-GCC-9.3.0.eb)
[OpenMPI-4.0.3-gcccuda-2020a.eb](https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.0.3-gcccuda-2020a.eb)
[OpenMPI-4.0.3-iccifort-2020.1.217.eb](https://github.com/easybuilders/easybuild-easyconfigs/blob/develop/easybuild/easyconfigs/o/OpenMPI/OpenMPI-4.0.3-iccifort-2020.1.217.eb)
all of these map to the same module file but are clearly not the same, this lack of uniqueness is a big problem as EB will see them all as installed once the module file is created.

It seems that the MNS is by design focussed on `foss` alone which means it is heavily exposed to problems that may occur when mixing different toolchains (for example fortran module incompatability between Intel/GCC). Even at `foss` level the MNS relies on there being no shadowing of software with different toolchains (for example, Python with `GCCcore` and the same version with `GCC`), this is currently true in recent releases but there is no guarantee that sites themselves respect this.